strawberryfields/backends/tfbackend/circuit.py

```diff
@@ -399,150 +364,126 @@ def prepare_multimode(self, state, modes, input_state_is_pure=False):
-            )
-            self._update_state(new_state)
+        k = tf.cast(kappa, ops.def_type)
+        k = self._maybe_batch(k)
+        new_state = ops.cross_kerr_interaction(
+            k, mode1, mode2, self._state, self._cutoff_dim, self._state_is_pure, self._batched
+        )
+        self._update_state(new_state)
 
     def cubic_phase(self, gamma, mode):
         """
         Apply the cubic phase operator to the specified mode.
         """
-        with self._graph.as_default():
-            g = tf.cast(gamma, ops.def_type)
-            g = self._maybe_batch(g)
-            new_state = ops.cubic_phase(
-                g,
-                mode,
-                self._state,
-                self._cutoff_dim,
-                self._hbar,
-                self._state_is_pure,
-                self._batched,
-            )
-            self._update_state(new_state)
+        g = tf.cast(gamma, ops.def_type)
+        g = self._maybe_batch(g)
+        new_state = ops.cubic_phase(
+            g, mode, self._state, self._cutoff_dim, self._hbar, self._state_is_pure, self._batched
+        )
+        self._update_state(new_state)
 
     def loss(self, T, mode):
         """
         Apply a loss channel  to the specified mode.
         """
-        with self._graph.as_default():
-            # in_modes = self._state
-            T = tf.cast(T, ops.def_type)
-            T = self._maybe_batch(T)
-            new_state = ops.loss_channel(
-                T, mode, self._state, self._cutoff_dim, self._state_is_pure, self._batched
-            )
-            self._update_state(new_state)
-            self._state_is_pure = False  # loss output always in mixed state representation
+        # in_modes = self._state
+        T = tf.cast(T, ops.def_type)
+        T = self._maybe_batch(T)
+        new_state = ops.loss_channel(
+            T, mode, self._state, self._cutoff_dim, self._state_is_pure, self._batched
+        )
+        self._update_state(new_state)
+        self._state_is_pure = False  # loss output always in mixed state representation
 
     def vacuum_element(self):
         """Compute the fidelity with the (multi-mode) vacuum state."""
-        with self._graph.as_default():
-            if self._batched:
-                vac_component = tf.reshape(self._state, [self._batch_size, -1])[:, 0]
-            else:
-                vac_component = tf.reshape(self._state, [-1])[0]
+        if self._batched:
+            vac_component = tf.reshape(self._state, [self._batch_size, -1])[:, 0]
+        else:
+            vac_component = tf.reshape(self._state, [-1])[0]
 
-            if self._state_is_pure:
-                return tf.abs(vac_component) ** 2
-            return vac_component
+        if self._state_is_pure:
+            return tf.abs(vac_component) ** 2
+        return vac_component
 
     def measure_fock(self, modes, select=None, **kwargs):
```

Comment:
This function is almost 200 lines long, it might be worth splitting it up for readability. The first ~50 lines are validation checks for example. Out of scope to split up the entire thing, just leaving it (maybe we could open an issue/TODO?).

Reply:
:grin:

Reply:
Since this is an existing function (i.e., there was no degradation), I saw we leave this for a separate PR.